google/cloud/bigtable/_mutate_rows.py

```diff
@@ -0,0 +1,220 @@
+        is_retryable_predicate: Callable[[Exception], bool],
+    ):
+        """
+        Helper class for managing saved state between mutate_rows attempts.
+
+        Args:
+          - gapic_fn: the function to call to trigger a mutate_rows rpc
+          - mutations: the list of mutations to send to the server
+          - timeout_generator: an iterator that yields values to use for the rpc timeout
+          - is_retryable_predicate: a function that returns True if an exception is retryable
+              should be the same predicate used by the retry wrapper
+        """
+        self.gapic_fn = gapic_fn
+        self.mutations = mutations
+        self.remaining_indices = list(range(len(mutations)))
+        self.timeout_generator = timeout_generator
+        self.is_retryable = is_retryable_predicate
+        self.errors: dict[int, list[Exception]] = {}
+
+    async def run_attempt(self):
+        """
+        Run a single attempt of the mutate_rows rpc.
+
+        Raises:
+          - _MutateRowsIncomplete: if there are failed mutations eligible for
+              retry after the attempt is complete
+          - GoogleAPICallError: if the gapic rpc fails
+        """
+        request_entries = [
+            self.mutations[idx]._to_dict() for idx in self.remaining_indices
+        ]
+        if not request_entries:
+            # no more mutations. return early
+            return
+        new_remaining_indices: list[int] = []
+        # make gapic request
+        try:
+            result_generator = await self.gapic_fn(
+                timeout=next(self.timeout_generator),
+                entries=request_entries,
+            )
+            async for result_list in result_generator:
+                for result in result_list.entries:
+                    # convert sub-request index to global index
+                    orig_idx = self.remaining_indices[result.index]
+                    entry_error = core_exceptions.from_grpc_status(
+                        result.status.code,
+                        result.status.message,
+                        details=result.status.details,
+                    )
+                    if result.status.code == 0:
+                        continue
+                    else:
+                        self.append_error(orig_idx, entry_error, new_remaining_indices)
+        except Exception as exc:
+            # add this exception to list for each active mutation
+            for idx in self.remaining_indices:
+                self.append_error(idx, exc, new_remaining_indices)
+                # bubble up exception to be handled by retry wrapper
+                raise
+        finally:
+            self.remaining_indices = new_remaining_indices
+        # check if attempt succeeded, or needs to be retried
+        if self.remaining_indices:
+            # unfinished work; raise exception to trigger retry
+            raise _MutateRowsIncomplete
+
+    def append_error(
```

Comment:
should this be prefixed with _ since its only called internally by _MutateRowsAttemptContext?

Reply:
Also please rename this to something thats a bit more broadly scoped....I was very surprised to see this messing with retry indices. Maybe _handle_entry_error?

Also I think it might be nicer to capture a local copy of self.remaining_indices before sending the RPC, set a new empty list on self for remaining_indices and have this method modify that as opposed to having this method take an output arg

Reply:
I ended up doing some refactoring to move the attempt method and the operation start function into the same class, to mirror what we're doing on the read_rows side. I think that makes things a bit cleaner, but let me know what you think